docs/storage-backend/inmemorybackend.md

```diff
@@ -16,18 +16,58 @@ graph = JanusGraphFactory.build().set('storage.backend', 'inmemory').open()
 
-The in-memory storage backend was primarily developed to simplify
+The in-memory storage backend was initially developed to simplify
 testing (for those tests that do not require persistence) and graph
-exploration. The in-memory storage backend is NOT meant for production
-use, large graphs, or high performance use cases. The in-memory storage
-backend is not performance or memory optimized. All data is stored in
-the heap space allocated to the Java virtual machine.
+exploration. Automated testing and ad-hoc prototyping remains its main purpose within Janusgraph project.
+
+Production:
+
+The initial test-only implementation was further evolved to use a more compact
+ in-memory representation, which made it suitable for production use in certain scenarios, such as 
+ Janusgraph engine being embedded into an application process, or spawned dynamically on demand, where:
+- Ease of setup/configuration/maintenance/start up is important (just run Janusgraph from distribution jar, no management of clusters for backend DBs etc) 
+- Loss of data due to _unexpected_ death of host process is acceptable (the backend provides a simple mechanism for making fast snapshots to handle _expected_ restarts)
+- Size of the graph data makes it possible to host it in a single JVM process (i.e. a few tens of Gigabytes max, unless you use a specialized JVM and hardware)
+- Higher performance is required, but no expertise/resources available to tune more complex backends. Due to its memory-only nature, in-memory backend typically performs faster than disk-based ones, in queries using simple indices
+and in graph modifications. However it is not specifically optimized for performance, and does not support advanced indexing functionality. 
+
+Limitations:
+
+- Obviously the scalability is limited to the heap size of a single JVM, and no transparent resilience to failures is offered
+- The backend offers store-level locking only, whereas a Janusgraph transaction typically changes multiple stores (e.g. vertex store and index store).
+This means that in scenarios where data is modified in parallel transactions, care should be taken on application level to avoid conflicting updates.
+At a high level, this means that you can modify unrelated parts of the graph in parallel, but you should avoid changing the same vertex or its edges in parallel
+transactions.
+- The backend does not guarantee a clean rollback once commit has started. Chances of a failure in the middle of commit to an in-memory data structure are low,
+however this can happen - e.g. when a large heap nears saturation and the GC pause exceeds configured backend timeout.
+- The data layout used by the backend can theoretically be susceptible to fragmentation in certain scenarios
+ (with a lot of add/delete operations), thus reducing the amount of useful data that can be stored in a heap
+  of specified size. The backend provides simple mechanisms to report fragmentation and defragment the storage if required.
```

Comment:
There currently is no support for activating these via configuration - one would have to have a little dig in the code to see how it can be invoked when needed. As and if some common patterns of how it can be applied for generic use case emerge - these patterns can be made configurable. For now, it can just be ignored, unless for a specific use case there is a suspicion that fragmentation might be a problem.